src/analyses/variable-sensitivity/monotonic_change.cpp

```diff
@@ -0,0 +1,331 @@
+  : abstract_value_objectt(t, tp, bttm),
+    monotonicity_value((tp || bttm) ? top_or_bottom : initial_value)
+{
+}
+
+monotonic_changet::monotonic_changet(
+  const exprt &e,
+  const abstract_environmentt &environment,
+  const namespacet &ns)
+  : abstract_value_objectt(e.type(), false, false), monotonicity_value(constant)
+{
+}
+
+index_range_implementation_ptrt
+monotonic_changet::index_range_implementation(const namespacet &ns) const
+{
+  return make_empty_index_range();
+}
+
+value_range_implementation_ptrt
+monotonic_changet::value_range_implementation() const
+{
+  return make_single_value_range(shared_from_this());
+}
+
+constant_interval_exprt monotonic_changet::to_interval() const
+{
+  return constant_interval_exprt(type());
+}
+
+void monotonic_changet::set_top_internal()
+{
+  monotonicity_value = top_or_bottom;
+}
+
+void monotonic_changet::output(
+  std::ostream &out,
+  const ai_baset &ai,
+  const namespacet &ns) const
+{
+  if(is_top() || is_bottom())
+    abstract_objectt::output(out, ai, ns);
+  else
+  {
+    switch(monotonicity_value)
+    {
+    case increase:
+      out << "Monotonically increasing";
+      break;
+    case decrease:
+      out << "Monotonically decreasing";
+      break;
+    case constant:
+      out << "Staying constant";
+      break;
+    case uninitialized:
+      out << "Declared but uninitialized";
+      break;
+    case top_or_bottom:
+    default:
+      out << "This should be unreachable";
+    }
+  }
+}
+
+abstract_object_pointert monotonic_changet::merge_with_value(
+  const abstract_value_pointert &other,
+  const widen_modet &widen_mode) const
+{
+  auto other_monotonic_change =
+    std::dynamic_pointer_cast<const monotonic_changet>(other);
+  if(other_monotonic_change != nullptr)
+    return merge_with_monotonic_change(other_monotonic_change);
+  else
+    return abstract_objectt::merge(other, widen_mode);
+}
+
+// It is weird if we ever try to merge uninitialized with increase, constant, or
+// decrease. In such a pathological case, we print out a message on the standard
+// output and simply return the top.
+abstract_object_pointert monotonic_changet::merge_with_monotonic_change(
+  const sharing_ptrt<const monotonic_changet> &other) const
+{
+  if(this->is_bottom() || other->is_top())
+    return std::make_shared<const monotonic_changet>(*other);
+  if(this->is_top() || other->is_bottom())
+    return shared_from_this();
+
+  if(this->monotonicity_value == increase)
+  {
+    switch(other->monotonicity_value)
+    {
+    case increase:
+    case constant:
+      return std::make_shared<monotonic_changet>(
+        this->type(), false, false, increase);
+    case decrease:
+      return std::make_shared<monotonic_changet>(
+        this->type(), true, false); // top
+    case uninitialized:
+      std::cout << "We are trying to merge increase with uninitialized\n";
+      break;
+    case top_or_bottom:
+    default:
+      std::cout << "This should be unreachable\n";
```

Comment:
Please use:
```suggestion
      UNREACHABLE;
```

(at several other places below as well)

Reply:
Thanks for the suggestion.